bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/ValueUtils.java

```diff
@@ -57,19 +75,135 @@ public class ValueUtils {
+                populateInitialValuesWithNoStrand(recordValue, latch, defaultValues);
+                latch.await();
+            } catch (InterruptedException e) {
+                throw ErrorCreator.createError(
+                        StringUtils.fromString("error occurred when populating default values"), e);
+            }
+        } else {
+            for (Map.Entry<String, BFunctionPointer<Object, ?>> field : defaultValues.entrySet()) {
+                recordValue.populateInitialValue(StringUtils.fromString(field.getKey()),
+                        field.getValue().call(new Object[]{strand}));
+            }
+        }
+        return recordValue;
+    }
+
+    private static Map<String, BFunctionPointer<Object, ?>> getNonProvidedDefaultValues
+            (Map<String, BFunctionPointer<Object, ?>> defaultValues, Set<String> providedFields) {
+        return defaultValues.entrySet().stream()
+                .filter(entry -> !providedFields.contains(entry.getKey()))
+                .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+    }
+
+    private static void populateInitialValuesWithNoStrand(BMap<BString, Object> recordValue, CountDownLatch latch,
+                                                          Map<String, BFunctionPointer<Object, ?>> defaultValues) {
+        String[] fields = defaultValues.keySet().toArray(new String[0]);
+        invokeFPAsyncIterativelyWithNoStrand(recordValue, defaultValues, fields, "default",
+                Scheduler.getDaemonStrand().getMetadata(), defaultValues.size(), o -> {
+        }, Scheduler.getDaemonStrand().scheduler, latch);
+    }
+
+    public static void invokeFPAsyncIterativelyWithNoStrand(BMap<BString, Object> recordValue,
+                                                            Map<String, BFunctionPointer<Object, ?>> defaultValues,
+                                                            String[] fields, String strandName, StrandMetadata metadata,
+                                                            int noOfIterations, Consumer<Object> futureResultConsumer,
+                                                            Scheduler scheduler, CountDownLatch latch) {
+        if (noOfIterations <= 0) {
+            return;
+        }
+        AtomicInteger callCount = new AtomicInteger(0);
+        scheduleNextFunction(recordValue, defaultValues, fields, strandName, metadata, noOfIterations, callCount,
+                futureResultConsumer, scheduler, latch);
+    }
+
+    private static void scheduleNextFunction(BMap<BString, Object> recordValue,
+                                             Map<String, BFunctionPointer<Object, ?>> defaultValues, String[] fields,
+                                             String strandName, StrandMetadata metadata, int noOfIterations,
+                                             AtomicInteger callCount, Consumer<Object> futureResultConsumer,
```

Comment:
Why do we need `AtomicInteger` for `callCount`? Can you see a way that this can get changed by multiple threads simultaneously?

Reply:
We need the `callCount` to be an `AtomicInteger` because it is getting updated inside the `notifySuccess` method of the callback. It will fail at compilation with the following error if we change it to `int`.
```
Variable 'callCount' is accessed from within inner class, needs to be final or effectively final
```